lib/srv/db/mysql/gcp.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package mysql
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/cloud/gcp"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+)
+
+func isDBUserGCPServiceAccount(dbUser string) bool {
+	if strings.Contains(dbUser, "@") {
+		switch {
+		// Example: mysql-iam-user@my-project-id.iam
+		// This format is used to align with PostgreSQL.
+		case strings.HasSuffix(dbUser, ".iam"):
+			return true
+		// Example: mysql-iam-user@my-project-id..gserviceaccount.com
+		case strings.HasSuffix(dbUser, ".iam.gserviceaccount.com"):
+			return true
+		}
+	}
+	return false
+}
+
+func gcpServiceAccountToDatabaseUser(serviceAccountName string) string {
+	user, _, _ := strings.Cut(serviceAccountName, "@")
+	return user
+}
+
+func databaseUserToGCPServiceAccount(sessionCtx *common.Session) string {
+	return fmt.Sprintf("%s@%s.iam.gserviceaccount.com", sessionCtx.DatabaseUser, sessionCtx.Database.GetGCP().ProjectID)
+}
+
+func (e *Engine) getGCPUserAndPassword(ctx context.Context, sessionCtx *common.Session, gcpClient gcp.SQLAdminClient) (string, string, error) {
+	// If `--db-user` is an service account email, use IAM Auth.
+	if isDBUserGCPServiceAccount(sessionCtx.DatabaseUser) {
+		return e.getGCPIAMUserAndPassword(ctx, sessionCtx)
+	}
+
+	// Get user info to decide how to authenticate.
+	user, err := gcpClient.GetUser(ctx, sessionCtx.Database, sessionCtx.DatabaseUser)
+	if err != nil {
+		// GetUser permission is new for IAM auth. If no permission, assume legacy password user.
+		if trace.IsAccessDenied(err) {
+			return e.getGCPPasswordUserAndPassword(ctx, sessionCtx)
```

Comment:
can we wrap the error using trace.Wrap ?

Reply:
i believe as long as the caller function returns a trace it should be fine. However, I refactored those functions to only return password and error now and errors are now wrapped.